### PR TITLE
Native POSIX Flash support

### PR DESCRIPTION
**Pull request by vanwinkeljan**
This PR adds flash support for the Native POSIX target, the flash data is placed on a binary file on the host system. ~~Further the PR provides the possibility to access FS partition on the flash by means of FUSE file system mounted on a host directory~~

Following features were included in this PR either to make it functional or to test it:
 * DTS support for Native POSIX
 * Updated test cases that require a flash but where not covered by QEMU
 * ~~Updated sanitycheck run directory~~
 * ~~Extended FS API to read mount points~~
 * ~~Added FS Shell Sample~~

This PR cover flash part mentioned in  issue #6044

FUSE support will be moved to separate PR
Moved sanitycheck changes to PR #16149

### CODE REVIEW

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [inline boards/posix/native_posix/doc/board.rst]: Shouldn't this commit then change the previous section about flash access, to mention this is also an option?

**Review by aescolar — CHANGES_REQUESTED** [review]
This will take a while to review.. Some minor things after a quick look thru

File: boards/posix/native_posix/native_posix.dts
Comment (vanwinkeljan) [reply boards/posix/native_posix/native_posix.dts]: This is not required as the DTS is only used to generate macro defines.
So as long you don't use the defines there is no problem.

File: boards/posix/native_posix/Kconfig
Comment (vanwinkeljan) [reply boards/posix/native_posix/Kconfig]: Nice catch, possibly some auto correct

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [inline boards/posix/native_posix/doc/board.rst]: typo: "option command line option" (?)

File: boards/posix/native_posix/fuse_flash.c
Comment (aescolar) [inline boards/posix/native_posix/fuse_flash.c]: After a quick look it would seem this file is not really part of the native_posix board, but could be shared by any posix arch based board, and could therefore have been anywhere else(?). If that is the case maybe it would be nicer to place it with the flash driver itself(?)
(The Kconfig option and cmake paragraph would go with it)

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [inline boards/posix/native_posix/doc/board.rst]: It feels like a lot more documentation could help the user.
What size is that file? is the file content just the raw binary dump? Are pages emulated? a nice drawing maybe?

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [inline boards/posix/native_posix/doc/board.rst]: nit: is present+,+

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [inline boards/posix/native_posix/doc/board.rst]: Some more documentation of this would not harm.. Right now the only documentation is this paragraph and the code (comment free ;) )

File: boards/posix/native_posix/fuse_flash.c
Comment (aescolar) [inline boards/posix/native_posix/fuse_flash.c]: I have no experience with the FUSE libray.. but seeing all this code I really wonder how stable its API is..

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [inline boards/posix/native_posix/doc/board.rst]: any version in particular you tested with and will maintain compatibility with?

File: dts/bindings/posix/zephyr,posix.yaml
Comment (aescolar) [inline dts/bindings/posix/zephyr,posix.yaml]: "of +a+ POSIX board" maybe?

File: subsys/settings/src/settings_file.c
Comment (aescolar) [inline subsys/settings/src/settings_file.c]: commit in master now..

File: tests/subsys/dfu/mcuboot/src/main.c
Comment (aescolar) [inline tests/subsys/dfu/mcuboot/src/main.c]: This file changes do not seem to belong in this commit..

**Review by aescolar — CHANGES_REQUESTED** [review]
a few more

File: boards/posix/native_posix/fuse_flash.c
Comment (vanwinkeljan) [reply boards/posix/native_posix/fuse_flash.c]: I do understand your concern but actually there is a lot of backwards compatibility in the library.
You may have notice the line ```#define FUSE_USE_VERSION 26``` this actually needs to be declared before including the fuse header file and specifies the API version used.

File: boards/posix/native_posix/doc/board.rst
Comment (vanwinkeljan) [reply boards/posix/native_posix/doc/board.rst]: > Any version in particular you tested with and will maintain compatibility with?

The minimum version is 2.6 and this is enforced by the code (```define FUSE_USE_VERSION 26```)

> So if some part does require something in particular, it would be nice to document it clearly.

The only restriction is that a UNIX like OS (Linux, MacOS X, free BSD, ...) is used with a minimal FUSE version of 2.6. Note that Windows/WSL is not supported but FUSE is highly requested because of it widespread usage (e.g. GlusterFs).

File: boards/posix/native_posix/fuse_flash.c
Comment (vanwinkeljan) [reply boards/posix/native_posix/fuse_flash.c]: Good point but would it not be better to move it then to subsys/fs?

File: tests/subsys/dfu/mcuboot/src/main.c
Comment (vanwinkeljan) [reply tests/subsys/dfu/mcuboot/src/main.c]: WIll move this to a seperate PR

File: boards/posix/native_posix/doc/board.rst
Comment (dbkinder) [reply boards/posix/native_posix/doc/board.rst]: I agree with @aescolar.  More information about this file is needed.  Does it have to exist before starting, or will it be created (as an empty file)?  Can you read or write past the end of file, or is the file exactly the flash size specified somewhere and you'll get some memory fault if reading beyond the file/flash size limit?  What's the recommended way to create this flash.bin file?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: tests/subsys/dfu/mcuboot/src/main.c
Comment (vanwinkeljan) [reply tests/subsys/dfu/mcuboot/src/main.c]: Moved to PR #12599 , will rebase this PR once merged

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [reply boards/posix/native_posix/doc/board.rst]: Ok. Would it make sense to mention this requirement in the doc?

File: boards/posix/native_posix/fuse_flash.c
Comment (aescolar) [reply boards/posix/native_posix/fuse_flash.c]: I have no opinion, I think it would be best if one of the flash or fs maintainers would comment here.

File: drivers/flash/flash_native_posix.c
Comment (aescolar) [inline drivers/flash/flash_native_posix.c]: This is a bit more flimsy than a user may understand from the documentation. If for example I give `-flash=banana/patata.bin` I get a silent segfault (if banana doesn't exist, that is)
Should the documentation mention instead of just "location", something like "location in an existing folder"?

File: boards/posix/native_posix/fuse_flash.c
Comment (aescolar) [inline boards/posix/native_posix/fuse_flash.c]: to +the+ directory where to mount +the+ flash

File: samples/subsys/fs/shell/README.rst
Comment (aescolar) [inline samples/subsys/fs/shell/README.rst]: .. a build, make sure..

File: samples/subsys/fs/shell/README.rst
Comment (aescolar) [inline samples/subsys/fs/shell/README.rst]: I have no experience with the filesystem subsystem, but I can't even get to list an empty mount point.
```
uart:~$ fs mount nffs /nffs
Successfully mounted fs:/nffs
uart:~$ fs ls /nffs
Unable to open /nffs (err -22)
```
or to mess with it from the FUSE mount side

**Review by aescolar — DISMISSED** [review]
I have no experience or knowledge about DTS or the filesystem part, so somebody else will need to review those.
@galak Would you do the honors for the DTS part? (or both)

File: boards/posix/native_posix/doc/board.rst
Comment (vanwinkeljan) [reply boards/posix/native_posix/doc/board.rst]: I could add it but dont know it is worth the effort as Zephyr only support WSL as build environment but WSL does not support 32-bit applications and as suche the native posix board is not supported on windows. This is in fact already clearly stated in the native posix board documentation.

File: samples/subsys/fs/shell/README.rst
Comment (vanwinkeljan) [reply samples/subsys/fs/shell/README.rst]: It is not you, I did hit the same issue.
After looking through the FS code it seems that it is limitation in Zephyrs FS implementation it seems that current implementation expects the root slash.

So ```fs ls /nffs/``` should do the trick, btw I did put a workaround in the Fuse code to handle this case.

File: drivers/flash/flash_native_posix.c
Comment (vanwinkeljan) [reply drivers/flash/flash_native_posix.c]: Wil add an extra member (init_called) to the data struct as 0 is a valid fd

File: boards/posix/native_posix/doc/board.rst
Comment (aescolar) [reply boards/posix/native_posix/doc/board.rst]: I meant the 2.6 version requirement, not the WSL thing.

File: boards/posix/native_posix/fuse_flash.c
Comment (vanwinkeljan) [reply boards/posix/native_posix/fuse_flash.c]: ^^^ @nashif @ramakrishnapallala

File: drivers/flash/flash_native_posix.c
Comment (aescolar) [reply drivers/flash/flash_native_posix.c]: > if the user does not enable logging the error will go unnoticed.

Given that it is a POSIX arch driver you could throw a `posix_print_warning` (which always goes to stderr)

**Review by dbkinder — APPROVED** [review]
+1 for docs, thanks!

**Comment by jhedberg** [comment]
Any update on this PR?

**Comment by aescolar** [comment]
@galak @nvlsianpu : Could you please review the DTS and file system sides respectively?

File: dts/bindings/posix/zephyr,posix.yaml
Comment (galak) [inline dts/bindings/posix/zephyr,posix.yaml]: we typically haven't had yaml for such things, is this actually needed for anythingn?

**Review by galak — COMMENTED** [review]
This looks good, but might be worth breaking up into a few PRs.  Would be good to explain the sanitycheck change a bit more in the commit message (why run in that dir)?

**Comment by vanwinkeljan** [comment]
@galak How would like to see the split?

Something like:
1. Native posix flash driver (including DTS)
2. Test updates and sanity check update
3. FS API update
4. Fuse flash access support for native posix and fs shell sample

Btw I also start wondering if there is not to much overlap with  issue #7403/ pr #8254 for point 1 and with issue #15180 for point 2. (@nvlsianpu @kapi-no @lemrey any input?)

**Comment by galak** [comment]
> @galak How would like to see the split?
> 
> Something like:
> 
> 1. Native posix flash driver (including DTS)
> 2. Test updates and sanity check update
> 3. FS API update
> 4. Fuse flash access support for native posix and fs shell sample
> 
> Btw I also start wondering if there is not to much overlap with issue #7403/ pr #8254 for point 1 and with issue #15180 for point 2. (@nvlsianpu @kapi-no @lemrey any input?)

I would have #1 & #2 as one PR.  Than #3 & #4 as a second PR.

**Comment by carlescufi** [comment]
@vanwinkeljan We have merged #8254, but I believe this is substantially different with distinct advantages for native posix.

File: boards/posix/native_posix/doc/index.rst
Comment (galak) [inline boards/posix/native_posix/doc/index.rst]: can you clarify, a 'flash device is present' ?

Comment (vanwinkeljan) [reply boards/posix/native_posix/doc/index.rst]: I will rephrase it to:
```In case flash driver support is enabled```

**Comment by vanwinkeljan** [comment]
@galak 
> I would have #1 & #2 as one PR. Than #3 & #4 as a second PR.

I will try to take care of this over the weekend

**Comment by vanwinkeljan** [comment]
@carlescufi @jhedberg 
> I believe this is substantially different with distinct advantages for native posix. 

My idea was to extend the simulated flash to have support for multiple backends, a RAM backend and a binary file backend. I did once a try by hacking the code of this PR in to the PR of the simulated flash driver but did run into some troubles with stats tracking which only supports a limited number of pages. So I will keep the native posix flash driver for now.

> In particular, and after a quick discussion with @jhedberg, this implementation would allow us to obtain a file that reflects the contents of a flash initialized with data.

This is actually why I implemented the fuse access, it would allow to create a file system with real content that can be used for regression testing.

**Comment by lemrey** [comment]
Like @carlescufi said, this PR brings in a new feature specifically for native POSIX, compared to the flash simulator in https://github.com/zephyrproject-rtos/zephyr/pull/8254, since it `mmaps` a file on the host so the simulated flash is persistent. Once memory mapped the flash is programmed as RAM just like in the flash simulator so perhaps, after https://github.com/zephyrproject-rtos/zephyr/pull/15663/ has gone in, it could be worth trying again to extend the simulator to support host files on POSIX, @vanwinkeljan?

**Review by aescolar — DISMISSED** [review]
This old comment still applies
https://github.com/zephyrproject-rtos/zephyr/pull/12503#discussion_r248321634

**Comment by vanwinkeljan** [comment]
Splitted PR in two, this PR only contains native posix flash implementation (including DTS support for native posix) and updates of test cases.

The second PR contains the FUSE based host access to the flash device.

File: dts/bindings/posix/zephyr,posix.yaml
Comment (vanwinkeljan) [reply dts/bindings/posix/zephyr,posix.yaml]: No it is not needed, file is removed

Comment (vanwinkeljan) [reply dts/bindings/posix/zephyr,posix.yaml]: Update: looks like it is needed for the bsim board else dts extraction script will fail because not a single binding is found.

https://github.com/zephyrproject-rtos/zephyr/blob/57629724c2b28acd857747cd232c5a975fe55591/scripts/dts/extract_dts_includes.py#L377

Comment (vanwinkeljan) [reply dts/bindings/posix/zephyr,posix.yaml]: Removed DTS support for nrf52 bsim support as such the unnecessary bindings file could be removed

**Review by galak — DISMISSED** [review]
Sorry if I didn't mention this before, but can you pull the 'sanitycheck: Execute binary in output directory' commit into its own PR.

**Comment by galak** [comment]
LGTM.

**Comment by vanwinkeljan** [comment]
@galak 

> Sorry if I didn't mention this before, but can you pull the 'sanitycheck: Execute binary in output directory' commit into its own PR.

Sorry my misunderstanding; I just split the PR in two, flash support and fuse access support.
I will move the sanitycheck commit in another PR, of course this PR will be blocked until it is merged and the commit will stay visible in this PR.

**Review by aescolar — COMMENTED** [review]
No complaints from me (note I did not retest it). But again, somebody who knows more about the flash side of things needs to give his opinion.
Just one request: Please add yourself also as codeowner to `drivers/flash/*native_posix*`
(Otherwise you could be sure I would be adding you to any review or bug for these files :) )

**Comment by vanwinkeljan** [comment]
> Just one request: Please add yourself also as codeowner to drivers/flash/*native_posix*

Done
 
>Sorry if I didn't mention this before, but can you pull the 'sanitycheck: Execute binary in output directory' commit into its own PR.

Done

**Comment by vanwinkeljan** [comment]
Added DNM label as this PR depends on #16149

File: CODEOWNERS
Comment (aescolar) [inline CODEOWNERS]: first '/' missing?
(I already get all `*native_posix*` drivers, so no need to have me in the line)

**Review by nashif — APPROVED** [review]

**Review by aescolar — APPROVED** [review]
I guess it is unlikely to get more reviews, so I think we can get this in, there is nothing wrong to me. (There would be just 1 nit: The 2nd UART being described in DTS for consistency, but it can come later)

**Pull request merged**
